src/c-writer.cc

```diff
@@ -0,0 +1,2437 @@
+        const Var& var = cast<CallExpr>(&expr)->var;
+        const Func& func = *module_->GetFunc(var);
+        Index num_params = func.GetNumParams();
+        Index num_results = func.GetNumResults();
+        assert(type_stack_.size() >= num_params);
+        if (num_results > 0) {
+          assert(num_results == 1);
+          Write(StackVar(num_params - 1, func.GetResultType(0)), " = ");
+        }
+
+        Write(GlobalVar(var), "(");
+        for (Index i = 0; i < num_params; ++i) {
+          if (i != 0)
+            Write(", ");
+          Write(StackVar(num_params - i - 1));
+        }
+        Write(");", Newline());
+        DropTypes(num_params);
+        PushTypes(func.decl.sig.result_types);
+        break;
+      }
+
+      case ExprType::CallIndirect: {
+        const FuncDeclaration& decl = cast<CallIndirectExpr>(&expr)->decl;
+        Index num_params = decl.GetNumParams();
+        Index num_results = decl.GetNumResults();
+        assert(type_stack_.size() > num_params);
+        if (num_results > 0) {
+          assert(num_results == 1);
+          Write(StackVar(num_params, decl.GetResultType(0)), " = ");
+        }
+
+        assert(module_->tables.size() == 1);
+        const Table* table = module_->tables[0];
+
+        assert(decl.has_func_type);
+        Index func_type_index = module_->GetFuncTypeIndex(decl.type_var);
+
+        Write("CALL_INDIRECT(", ExternalRef(table->name), ", ");
+        WriteFuncDeclaration(decl, "(*)");
+        Write(", ", func_type_index, ", ", StackVar(0));
+        for (Index i = 0; i < num_params; ++i) {
+          Write(", ", StackVar(num_params - i));
+        }
+        Write(");", Newline());
+        DropTypes(num_params + 1);
+        PushTypes(decl.sig.result_types);
+        break;
+      }
+
+      case ExprType::Compare:
+        Write(*cast<CompareExpr>(&expr));
+        break;
+
+      case ExprType::Const: {
+        const Const& const_ = cast<ConstExpr>(&expr)->const_;
+        PushType(const_.type);
+        Write(StackVar(0), " = ", const_, ";", Newline());
+        break;
+      }
+
+      case ExprType::Convert:
+        Write(*cast<ConvertExpr>(&expr));
+        break;
+
+      case ExprType::CurrentMemory: {
+        assert(module_->memories.size() == 1);
+        Memory* memory = module_->memories[0];
+
+        PushType(Type::I32);
+        Write(StackVar(0), " = ", ExternalRef(memory->name), ".pages;",
+              Newline());
+        break;
+      }
+
+      case ExprType::Drop:
+        DropTypes(1);
+        break;
+
+      case ExprType::GetGlobal: {
+        const Var& var = cast<GetGlobalExpr>(&expr)->var;
+        PushType(module_->GetGlobal(var)->type);
+        Write(StackVar(0), " = ", GlobalVar(var), ";", Newline());
+        break;
+      }
+
+      case ExprType::GetLocal: {
+        const Var& var = cast<GetLocalExpr>(&expr)->var;
+        PushType(func_->GetLocalType(var));
+        Write(StackVar(0), " = ", var, ";", Newline());
```

Comment:
That there is something going on with var naming

```
(module
  (func $addTwo (param i32) (result i32)
    (local i32)
    get_local 1
  )
)
```

gets compiled as

```c
static u32 addTwo(u32 _) {
  u32 __0 = 0;
  FUNC_PROLOGUE;
  u32 i0;
  i0 = _;
  FUNC_EPILOGUE;
  return i0;
}
```

it should return `__0` (0) however it returns `_` (arg passed in the func).